taichi/codegen/cuda/codegen_cuda.cpp

```diff
@@ -161,6 +162,39 @@ class TaskCodeGenCUDA : public TaskCodeGenLLVM {
     llvm_val[stmt] = create_print(formats, types, values);
   }
 
+  void visit(AllocaStmt *stmt) override {
+    // Override shared memory codegen logic for large shared memory
+    if (stmt->ret_type->is<TensorType>() && stmt->is_shared) {
+      auto tensor_type = stmt->ret_type->cast<TensorType>();
+      size_t shared_array_threshold = 49152;
+      size_t shared_array_bytes =
+          tensor_type->get_num_elements() *
+          data_type_size(tensor_type->get_element_type());
+      if (shared_array_bytes > shared_array_threshold) {
+        if (dynamic_shared_array_bytes > 0) {
+          TI_ERROR(
+              "Only one single large shared array instance is allowed in "
+              "current "
+              "version.")
+        }
+        // Clear tensor shape for dynamic shared memory.
+        tensor_type->set_shape(std::vector<int>({0}));
+        dynamic_shared_array_bytes += shared_array_bytes;
```

Comment:
OOC:
1. from the logic here we only allow creating one large shared array but why does `test_multiple_shared_array` also pass? Did I miss anything obvious? :P 
2. Looks like `dynamic_shared_array_bytes` will be incremented at most once, shall we use `=` instead of `+=` to make it more clear?